### PR TITLE
[WIP] group by

### PR DESCRIPTION
**Pull request by p0lyn0mial**

### CODE REVIEW

File: pkg/cmd/audit/aggregate.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+import (
+	"strings"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	auditv1 "k8s.io/apiserver/pkg/apis/audit/v1"
+	"k8s.io/apiserver/pkg/audit"
+)
+
+func GroupBy(paths []string,  events []*auditv1.Event) map[string][]*auditv1.Event {
```
Comment (p0lyn0mial) [inline pkg/cmd/audit/aggregate.go:11]: you can group by `user name` passing `[]string{"user.username"`
or by `verb, uri and user name` passing `[]string{"verb", "requestURI", "user.username"}`

File: pkg/cmd/audit/func.go [hunk 2]
```diff
@@ -0,0 +1,18 @@
+package audit
+
+import (
+	"sort"
+
+	auditv1 "k8s.io/apiserver/pkg/apis/audit/v1"
+)
+
+func SortByLen(events map[string][]*auditv1.Event) []string {
```
Comment (p0lyn0mial) [inline pkg/cmd/audit/func.go:9]: this sorts aggregated events by the number of entries

File: pkg/cmd/audit/audit_test.go [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+	}
+}
+
+func TestExampleTop5ByVerbURIAndUserName(t *testing.T) {
+	groupBy := []string{"verb", "requestURI", "user.username"}
+	events, err := GetEvents("./testdata/audit.log")
+	if err != nil {
+		t.Fatalf("unable to read test data, err %v", err)
+	}
+	aggregatedEvents := GroupBy(groupBy, events)
```
Comment (p0lyn0mial) [inline pkg/cmd/audit/audit_test.go:76]:
```
aggregatedEvents := GroupBy(groupBy, events)
sortedEvents := SortByLen(aggregatedEvents)
```

and this is how you could get top 5 by verb, uri and user name

File: pkg/cmd/audit/aggregate.go [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+	result := map[string][]*auditv1.Event{}
+	for _, event := range events {
+		unstructuredEvent := map[string]interface{}{}
+		{
+			asUnstructured := &unstructured.Unstructured{}
+			if err := audit.Scheme.Convert(event, asUnstructured, nil); err != nil {
+				panic(err)
```
Comment (sttts) [inline pkg/cmd/audit/aggregate.go:18]: ugly

File: pkg/cmd/audit/aggregate.go [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	auditv1 "k8s.io/apiserver/pkg/apis/audit/v1"
+	"k8s.io/apiserver/pkg/audit"
+)
+
+func GroupBy(paths []string,  events []*auditv1.Event) map[string][]*auditv1.Event {
+	result := map[string][]*auditv1.Event{}
+	for _, event := range events {
+		unstructuredEvent := map[string]interface{}{}
+		{
```
Comment (sttts) [inline pkg/cmd/audit/aggregate.go:15]: why this block?

File: pkg/cmd/audit/aggregate_test.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
+package audit
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
Comment (sttts) [inline pkg/cmd/audit/aggregate_test.go:4]: order

File: pkg/cmd/audit/query_test.go [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+)
+
+func TestDriveQuery(t *testing.T) {
+	res, err := audit.Read("./testdata/audit.log").GroupBy([]string{"user.username"}).SortBy("length").Run()
+	if err != nil {
+		t.Fatalf("unexpected err %v", err)
+	}
+
+	// TODO: validate res
+	_ = res
```
Comment (p0lyn0mial) [inline pkg/cmd/audit/query_test.go:15]: res is an array of

```
type Result struct {
	Key   string
	Group []*auditv1.Event
}
```

**Pull request closed without merging**